graphql-lint/src/lib.rs

```diff
@@ -0,0 +1,654 @@
+        } else {
+            CaseMatch::Incorrect { current, fix }
+        }
+    }
+
+    pub fn visit_field_argument(
+        &mut self,
+        parent_type: &TypeDefinition<'a>,
+        field: &FieldDefinition<'a>,
+        argument: &InputValueDefinition<'a>,
+    ) {
+        if let CaseMatch::Incorrect { current, fix } = Self::case_check(argument.name(), Case::Camel) {
+            self.diagnostics.push((
+                format!(
+                    "argument '{current}' on field '{}' on {} '{}' should be renamed to '{fix}'",
+                    field.name(),
+                    Self::type_definition_display(parent_type),
+                    parent_type.name()
+                ),
+                Severity::Warning,
+            ));
+        }
+    }
+
+    pub fn visit_directive_argument(
+        &mut self,
+        directive: &DirectiveDefinition<'a>,
+        argument: &InputValueDefinition<'a>,
+    ) {
+        if let CaseMatch::Incorrect { current, fix } = Self::case_check(argument.name(), Case::Camel) {
+            self.diagnostics.push((
+                format!(
+                    "argument '{current}' on directive '{}' should be renamed to '{fix}'",
+                    directive.name()
+                ),
+                Severity::Warning,
+            ));
+        }
+    }
+
+    pub fn visit_input_value(&mut self, parent: &TypeDefinition<'a>, value: &InputValueDefinition<'a>) {
+        if let CaseMatch::Incorrect { current, fix } = Self::case_check(value.name(), Case::Camel) {
+            self.diagnostics.push((
+                format!(
+                    "input value '{current}' on input '{}' should be renamed to '{fix}'",
+                    parent.name()
+                ),
+                Severity::Warning,
+            ));
+        }
+    }
+
+    fn type_definition_display(kind: &TypeDefinition<'a>) -> &'static str {
+        match kind {
+            TypeDefinition::Scalar(_) => "scalar",
+            TypeDefinition::Object(_) => "type",
+            TypeDefinition::Interface(_) => "interface",
+            TypeDefinition::Union(_) => "union",
+            TypeDefinition::Enum(_) => "enum",
+            TypeDefinition::InputObject(_) => "input",
+        }
+    }
+
+    pub fn visit_field(&mut self, parent: &TypeDefinition<'a>, field: &FieldDefinition<'a>) {
+        let field_name = field.name();
+
+        // ignore system fields
+        if field_name.starts_with("__") {
+            return;
+        }
+
+        if let CaseMatch::Incorrect { current, fix } = Self::case_check(field_name, Case::Camel) {
+            self.diagnostics.push((
+                format!(
+                    "field '{current}' on {} '{}' should be renamed to '{fix}'",
+                    Self::type_definition_display(parent),
+                    parent.name()
+                ),
+                Severity::Warning,
+            ));
+        }
+        match parent.name() {
+            "Query" => {
```

Comment:
The root types aren't _neccesarily_ named `Query`, `Mutation` & `Subscription`.  They are the default names, but might want to check if there's a schema definition and use the names from there if it exists

Reply:
That's true, I was following the lead of ESLint here (e.g. `FieldDefinition[parent.name.value=Subscription]`). We can make it more specific later if needed, although I think for our purposes they'll always be named that

Reply:
The issue with checking the schema definition is that I can't guarantee it'll show up before this check, so it'd require a second pass

Reply:
> The issue with checking the schema definition is that I can't guarantee it'll show up before this check, so it'd require a second pass

Sure but a `schema.definitions().find(|def| matches!(def, Definition::SchemaDefinition(_))` seems relatively easy to do.  Anyway, I'm not going to fight for this, so long as we're aware

Reply:
It's definitely the plan to do so, I'm not sure we want to do it for the MVP but I'll possibly add it in